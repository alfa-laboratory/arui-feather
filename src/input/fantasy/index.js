/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import './input.css';
import '../input_mode_link.css'; // TODO @teryaew: not changed?
import './input_theme_alfa-on-color.css';
import './input_theme_alfa-on-white.css';

export default from '../input';
