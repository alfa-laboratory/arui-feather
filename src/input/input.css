/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

@import "../vars.css";

:root {
    --top-scale-s: calc(strip(var(--font-size-xs)) / strip(var(--font-size-s)));
    --top-scale-m: calc(strip(var(--font-size-s)) / strip(var(--font-size-m)));
    --top-scale-l: calc(strip(var(--font-size-m)) / strip(var(--font-size-l)));
    --top-scale-xl: calc(
        strip(var(--font-size-l)) / strip(var(--font-size-xl))
    );
}

.input {
    display: inline-block;
    font-family: var(--font);
    font-weight: var(--font-weight-normal);

    &,
    *,
    *:before,
    *:after {
        box-sizing: border-box;
    }

    &__inner {
        display: inline-table;
        position: relative;
        width: 100%;
    }

    &__box {
        display: table-cell;
        position: relative;
        user-select: none;
        transition-duration: 250ms;
        transition-property: border-bottom-color, box-shadow, width,
            background-color;
        transition-timing-function: cubic-bezier(
            0.23,
            1,
            0.32,
            1
        ); /* TODO @teryaew: add custom easings to styles */
        -webkit-touch-callout: none;
    }

    &__input-wrapper {
        position: relative;
        display: block;
    }

    &__top {
        overflow: hidden;
        position: absolute;
        top: 0;
        right: 0;
        left: 0;
        font-size: var(--font-size-m);
        text-overflow: ellipsis;
        white-space: nowrap;
        transform: scale(1) translateY(0);
        transform-origin: 0 100%;
        transition-duration: 200ms;
        transition-property: color, transform, font-size;
        transition-timing-function: cubic-bezier(0.25, 0.1, 0.25, 1);
    }

    &__sub {
        display: table-caption;
        caption-side: bottom;
        padding-top: 5px;
        padding-right: var(--gap-s);
        white-space: normal;
    }

    &__control {
        display: inline-block;
        position: relative;
        width: 100%;

        /** Если использовать `height: 100%` поле схлопывается в Safari v10.1.2. **/
        min-height: 100%;
        margin: 0;
        padding: 0;
        outline: none;
        font: inherit;
        font-weight: var(--font-weight-normal);
        line-height: inherit;
        background: none;
        border: none;
        border-radius: 0;
        -webkit-appearance: none;
        -webkit-tap-highlight-color: rgba(0, 0, 0, 0);

        &:invalid {
            box-shadow: none;
        }

        &::-webkit-search-decoration {
            -webkit-appearance: none;
        }

        &::-webkit-search-cancel-button {
            display: none;
        }

        &::-webkit-input-placeholder {
            text-indent: 0;
        }

        &::-moz-placeholder {
            opacity: var(--opacity-active);
        }

        &::-ms-clear {
            display: none;
        }

        &[placeholder],
        &::placeholder {
            text-overflow: ellipsis;
            transition: opacity 0.2s ease-in-out, color 0.2s ease-in-out;
        }

        &:-webkit-autofill,
        &:-webkit-autofill:hover,
        &:-webkit-autofill:focus,
        &:-webkit-autofill:active {
            /* Hack from http://stackoverflow.com/a/29350537 */
            transition: background-color 999999s ease-in-out 0s;
        }
    }

    &_has-label {
        .input__control::placeholder {
            opacity: 0;
        }

        &.input_focused {
            .input__control::placeholder {
                opacity: var(--opacity-active);
            }
        }
    }

    &_has-clear &__control {
        padding-right: 0;
        text-overflow: ellipsis;
    }

    &__clear {
        opacity: var(--opacity-minor);
        position: absolute;
        top: 0;
        right: 0;
        margin: 0;
        padding: 0;
        cursor: pointer;
        outline: none;
        user-select: none;
        color: inherit;
        font: inherit;
        background: none;
        background-position: center center;
        background-repeat: no-repeat;
        border: none;
        transition: opacity 0.1s ease-out;

        &:hover {
            opacity: var(--opacity-active);
        }
    }

    &_has-icon &__icon {
        display: flex;
        align-items: center;
        justify-content: center;
        position: absolute;
        top: 0;
        right: 0;
        text-align: center;
        transition: all 200ms ease;
    }

    &_has-icon &__control {
        padding-right: 2px;
        text-overflow: ellipsis;
    }

    &_width_available {
        width: 100%;
    }

    &&_has-addons {
        /* Override table styles by flex */
        .input__box {
            display: flex;
            flex-direction: row;
            flex-grow: 1;
            justify-content: space-between;
            width: 100%;
            height: 100%;
        }

        .input__control {
            padding-right: var(--gap-s);
        }

        .input__input-wrapper:nth-child(2) {
            .input__control,
            .input__top {
                padding-right: 0;
                padding-left: var(--gap-s);
            }
        }

        /* If has left & right addons */
        .input__input-wrapper:nth-child(2):nth-last-child(2) {
            .input__control,
            .input__top {
                padding-right: var(--gap-s);
            }
        }
    }

    &__addons {
        display: flex;
        flex-shrink: 0; /* Avoid overflow cutting of addons in IE11 */
        align-items: center;
    }

    &_type_hidden {
        display: none !important;
    }

    /* reset for ios */
    &__control:disabled {
        opacity: var(--opacity-active);
    }
}

.input_focused {
    z-index: 1;
}

.input_disabled {
    &:before,
    .input__clear {
        display: none;
    }

    .input_has-value {
        &::placeholder {
            opacity: var(--opacity-minor);
        }
    }

    .input__box,
    .input__control {
        cursor: default;
    }
}

.input_type_password .input__control {
    text-overflow: clip;
}

.input_size_s {
    &.input_has-label {
        padding-top: 9px;
    }

    .input__top {
        line-height: var(--font-size-2xl);
    }

    &.input_focused,
    &.input_has-value {
        .input__top {
            transform: scale(var(--top-scale-s)) translateY(-22px);
            width: calc((2 - var(--top-scale-s)) * 100) %;
        }
    }

    .input__top,
    .input__box {
        font-size: var(--font-size-s);
    }

    .input__box {
        height: 30px;
    }

    .input__sub {
        font-size: var(--font-size-xs);
    }

    .input__control {
        padding: 7px 0;
    }

    &.input_has-clear .input__box,
    &.input_has-icon .input__box {
        padding-right: 30px;
    }

    &.input_has-icon .input__icon {
        width: 30px;
        height: 30px;
        line-height: 28px;
    }
}

.input_size_m {
    &.input_has-label {
        padding-top: 6px;
    }

    .input__top {
        line-height: var(--font-size-3xl);
    }

    &.input_focused,
    &.input_has-value {
        .input__top {
            transform: scale(var(--top-scale-m)) translateY(-30px);
            width: calc((2 - var(--top-scale-m)) * 100) %;
        }
    }

    .input__top,
    .input__box {
        font-size: var(--font-size-m);
    }

    .input__box {
        height: 40px;
    }

    .input__sub {
        font-size: var(--font-size-s);
    }

    .input__control {
        padding: 10px 0;
    }

    &.input_has-clear .input__box,
    &.input_has-icon .input__box {
        padding-right: 34px;
    }

    &.input_has-icon .input__icon {
        width: 34px;
        height: 40px;
        line-height: 38px;
    }
}

.input_size_l {
    &.input_has-label {
        padding-top: 14px;
    }

    .input__top {
        line-height: 50px;
    }

    &.input_focused,
    &.input_has-value {
        .input__top {
            transform: scale(var(--top-scale-l)) translateY(-40px);
            width: calc((2 - var(--top-scale-l)) * 100) %;
        }
    }

    .input__top,
    .input__box {
        font-size: var(--font-size-l);
    }

    .input__box {
        height: 50px;
    }

    .input__sub {
        font-size: var(--font-size-m);
    }

    .input__control {
        padding: 14px 0;
    }

    &.input_has-clear .input__box,
    &.input_has-icon .input__box {
        padding-right: 44px;
    }

    &.input_has-icon .input__icon {
        width: 44px;
        height: 50px;
        line-height: var(--font-size-4xl);
    }
}

.input_size_xl {
    &.input_has-label {
        padding-top: 13px;
    }

    .input__top {
        line-height: 60px;
    }

    &.input_focused,
    &.input_has-value {
        .input__top {
            transform: scale(var(--top-scale-xl)) translateY(-50px);
            width: calc((2 - var(--top-scale-xl)) * 100) %;
        }
    }

    .input__top,
    .input__box {
        font-size: var(--font-size-xl);
    }

    .input__box {
        height: 60px;
    }

    .input__sub {
        font-size: var(--font-size-l);
    }

    .input__control {
        padding: 17px 0;
    }

    &.input_has-clear .input__box,
    &.input_has-icon .input__box {
        padding-right: 52px;
    }

    &.input_has-icon .input__icon {
        width: 52px;
        height: 60px;
        line-height: 58px;
    }
}

/* IE fix for input with label and placeholder */
.input_has-label .input__control:-ms-input-placeholder {
    opacity: 0;
}

.input_view_filled {
    &.input_size_s {
        .input__top {
            line-height: 43px;
        }

        .input__box {
            height: 42px;
            padding: 0 var(--gap-xs);
        }

        .input__control {
            height: 42px;
        }

        .input__top,
        .input__box {
            font-size: var(--font-size-s);
        }

        &.input_has-label {
            .input__control {
                padding-top: 21px;
                padding-bottom: 7px;
            }
        }

        &.input_focused,
        &.input_has-value {
            .input__top {
                font-size: var(--font-size-xs);
                transform: translateY(-10px);
            }
        }

        &.input_has-icon,
        &.input_has-clear {
            .input__icon,
            .input__clear {
                top: 50%;
                width: 40px;
                height: 42px;
                margin-top: -21px;
            }

            .input__box {
                padding-right: var(--gap-3xl);
            }
        }

        &.input_has-addons {
            .input__box {
                height: 43px;
            }
        }
    }

    &.input_size_m {
        .input__top {
            line-height: 56px;
        }

        .input__box {
            height: 56px;
            padding: 0 var(--gap-s);
        }

        .input__control {
            min-height: 56px;
        }

        .input__top,
        .input__box {
            font-size: var(--font-size-m);
        }

        &.input_has-label {
            .input__control {
                padding-top: 26px;
                padding-bottom: var(--gap-xs);
            }
        }

        &.input_focused,
        &.input_has-value {
            .input__top {
                font-size: var(--font-size-s);
                transform: translateY(-12px);
            }
        }

        &.input_has-icon,
        &.input_has-clear {
            .input__icon,
            .input__clear {
                top: 50%;
                width: 48px;
                height: 48px;
                margin-top: var(--gap-xl-neg);
            }

            .input__box {
                padding-right: var(--gap-4xl);
            }
        }

        &.input_has-addons {
            .input__box {
                height: 56px;
            }
        }
    }

    .input__top {
        transform-origin: 0 50%;
    }

    .input__box {
        border-bottom-color: var(--color-dark-indigo-15);
        background-color: var(--color-dark-indigo-05);
        border-radius: 4px 4px 0 0;

        &:hover {
            background-color: var(--color-dark-indigo-10);
            border-color: var(--color-dark-indigo-60);
        }
    }

    &__sub {
        padding-top: var(--gap-2xs);
        padding-right: var(--gap-s);
        font-size: var(--font-size-s);
        line-height: var(--line-height-condensed);
    }

    .input__top,
    .input__sub {
        color: var(--color-dark-indigo-60);
    }

    .input__control {
        padding: var(--gap-xs) 0;
        color: var(--color-dark-indigo);

        &::placeholder {
            color: var(--color-dark-indigo-60);
        }

        /*
        &:-webkit-autofill,
        &:-webkit-autofill:hover,
        &:-webkit-autofill:focus,
        &:-webkit-autofill:active {
            -webkit-text-fill-color: var(--color-content-alfa-on-white) !important;
        }
        */
    }

    &.input_focused {
        .input__box {
            border-bottom-color: var(--color-dark-indigo);
            box-shadow: inset 0 -1px 0 var(--color-dark-indigo);
        }
    }

    &.input_has-label {
        padding-top: 0;

        .input__control::placeholder {
            color: transparent;
        }

        &.input_focused {
            .input__control::placeholder {
                color: var(--color-dark-indigo-60);
            }
        }
    }

    &.input_focused,
    &.input_has-value {
        .input__top {
            width: auto;
        }
    }

    &.input_has-addons {
        .input__input-wrapper {
            .input__control {
                padding-left: var(--gap-s);
                padding-right: var(--gap-s);
            }

            &:first-child {
                .input__control {
                    padding-left: 0;
                }
            }

            &:last-child {
                .input__control {
                    padding-right: 0;
                }
            }

            &:nth-child(2) {
                .input__control {
                    padding-right: var(--gap-s);
                }
            }
        }
    }

    &.input_disabled {
        .input__box {
            background-color: var(--color-dark-indigo-10);
            border-bottom-color: var(--color-dark-indigo-15);
            border-bottom-style: var(--border-style-control-disabled);
            box-shadow: none;
        }

        .input__control {
            color: var(--color-dark-indigo-60);
        }

        .input__top,
        .input__control::placeholder {
            color: var(--color-dark-indigo-30);
        }
    }

    &.input_invalid {
        .input__sub {
            color: var(--color-error);
        }

        .input__box {
            border-bottom-color: var(--color-error);
            box-shadow: inset 0 -1px 0 var(--color-error);
        }

        &.input_focused {
            .input__box {
                border-bottom-color: var(--color-error);
                box-shadow: inset 0 -1px 0 var(--color-error);
            }
        }
    }
}
