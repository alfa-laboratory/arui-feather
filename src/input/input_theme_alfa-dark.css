/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

@import "../vars.css";
@import "@alfalab/core-components-vars/colors-bluetint.css";

.input_theme_alfa-dark {
    .input__box {
        border-bottom-color: var(--color-dark-graphic-tertiary);
        border-bottom-style: solid;
        border-bottom-width: 1px;
    }

    &.input_view_filled .input__box,
    &.input_view_filled .input__box:hover {
        background-color: var(--color-dark-bg-secondary);
    }

    .input__box:hover {
        border-bottom-color: var(--color-dark-graphic-primary);
    }

    .input__control {
        color: var(--color-dark-text-primary);

        &::placeholder {
            color: var(--color-dark-text-secondary-transparent);
        }

        &:-webkit-autofill,
        &:-webkit-autofill:hover,
        &:-webkit-autofill:focus,
        &:-webkit-autofill:active {
            -webkit-text-fill-color: var(--color-dark-text-primary);
            caret-color: var(--color-dark-text-primary);
        }
    }

    .input__top,
    .input__sub {
        color: var(--color-dark-text-secondary-transparent);
    }

    &.input_has-label {
        .input__control::placeholder {
            color: transparent;
        }

        &.input_focused {
            .input__control::placeholder {
                color: var(--color-dark-text-secondary-transparent);
            }
        }
    }

    &.input_has-icon {
        .input__icon {
            color: var(--color-dark-text-primary);
        }
    }

    &.input_focused {
        .input__box {
            border-bottom-color: var(--color-dark-graphic-primary);
            box-shadow: inset 0 -1px 0 var(--color-dark-graphic-primary);
        }
    }

    &.input_disabled {
        .input__box {
            border-bottom-color: var(--color-dark-bg-secondary-inverted);
            border-bottom-style: var(--border-style-control-disabled);
            box-shadow: none;
        }

        &,
        .input__control {
            color: var(--color-light-text-secondary);

            &:-webkit-autofill,
            &:-webkit-autofill:hover,
            &:-webkit-autofill:focus,
            &:-webkit-autofill:active {
                -webkit-text-fill-color: var(--color-light-text-secondary);
            }
        }

        .input__control::placeholder,
        .input__top {
            color: var(--color-light-text-secondary);
        }

        .input__icon {
            opacity: var(--opacity-disabled);
        }
    }

    &.input_invalid {
        .input__sub {
            color: var(--color-dark-text-negative);
        }

        .input__box {
            border-bottom-color: var(--color-dark-graphic-negative);
        }

        &.input_focused {
            .input__box {
                border-bottom-color: var(--color-dark-graphic-negative);
                box-shadow: inset 0 -1px 0 var(--color-dark-graphic-negative);
            }
        }
    }
}
