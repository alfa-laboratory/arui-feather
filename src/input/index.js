/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import '../masked-input';
import '../popup';

import './input_theme_alfa-on-color.css';
import './input_theme_alfa-on-white.css';
import './input.css';

export default from './input';
