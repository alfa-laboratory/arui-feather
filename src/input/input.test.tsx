/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* eslint import/no-extraneous-dependencies: [2, {"devDependencies": true}] */

import React from 'react';
import { shallow, mount } from 'enzyme';

import { Input } from './input';
import MaskedInput from '../masked-input';
import IconSearch from '../icon/action/search';

import { SCROLL_TO_CORRECTION } from '../vars';

describe('input', () => {
    const originalWindowScrollTo = window.scrollTo;

    beforeEach(() => {
        window.scrollTo = jest.fn();
    });

    afterEach(() => {
        window.scrollTo = originalWindowScrollTo;
    });

    it('should unset class on input blur', () => {
        const input = shallow(<Input />);
        const controlNode = input.find('input');

        controlNode.simulate('focus');
        expect(input.props().className).toContain('input_focused');

        controlNode.simulate('blur');
        expect(input.props().className).not.toContain('input_focused');
    });

    it('should render with disabled class', () => {
        const input = shallow(<Input disabled={ true } />);

        expect(input.props().className).toContain('input_disabled');
    });

    it('should unset class on blur event', () => {
        const input = mount(<Input />);

        input.find('input').simulate('focus');

        expect(input.children().props().className).toContain('input_focused');

        input.find('input').simulate('blur');

        expect(input.children().props().className).not.toContain('input_focused');
    });

    it('should set caret to end of input on public focus method', () => {
        const value = 'test';
        const input = mount<Input>(<Input value={ value } />);

        input.instance().focus();

        expect(input.instance().getControl().selectionStart).toBe(value.length);
        expect(input.instance().getControl().selectionEnd).toBe(value.length);
    });

    it('should set class on focus event', () => {
        const input = mount(<Input />);

        input.find('input').simulate('focus');

        expect(input.children().props().className).toContain('input_focused');
    });

    it('should scroll window to element on public `scrollTo` method', () => {
        const input = mount<Input>(<Input />);
        const elemTopPosition = input.getDOMNode().getBoundingClientRect().top;
        const elemScrollTo = (elemTopPosition + window.pageYOffset) - SCROLL_TO_CORRECTION;

        input.instance().scrollTo();

        expect(window.scrollTo).toHaveBeenCalledWith(0, elemScrollTo);
    });

    it(
        'should not set selection when `setSelectionRange` method was called on input with type `email`',
        () => {
            const input = mount<Input>(<Input value='test' type='email' />);

            const setSelectionRange = jest.fn();

            input.instance().getControl().setSelectionRange = setSelectionRange;
            input.instance().focus();

            input.instance().setSelectionRange();
            expect(setSelectionRange).not.toHaveBeenCalled();
        }
    );

    it(
        'should set selection to all value when `setSelectionRange` method was called without parameters',
        () => {
            const input = mount<Input>(<Input value='test' />);
            const setSelectionRange = jest.fn();

            input.instance().getControl().setSelectionRange = setSelectionRange;

            input.instance().focus();

            input.instance().setSelectionRange();
            expect(setSelectionRange).toHaveBeenCalledWith(0, 'test'.length);
        }
    );

    it(
        'should set selection when setSelection range was called with parameters',
        () => {
            const input = mount<Input>(<Input value='test' />);
            const setSelectionRange = jest.fn();

            input.instance().getControl().setSelectionRange = setSelectionRange;

            input.instance().focus();

            input.instance().setSelectionRange(0, 2);
            expect(setSelectionRange).toHaveBeenCalledWith(0, 2);

            input.instance().setSelectionRange(2, 4);
            expect(setSelectionRange).toHaveBeenCalledWith(2, 4);
        }
    );

    it('should render without problems', () => {
        const input = shallow(<Input />);

        expect(input).toBeDefined();
        expect(input).toMatchSnapshot();
        expect(input.props().className).toContain('input');
    });

    it('should set focused className from props', () => {
        const input = shallow(<Input focused={ true } />);

        expect(input.props().className).toContain('input_focused');
    });

    it('should render with `label` from props', () => {
        const input = shallow(<Input label='Label' />);
        const topNode = input.find('.input__top');

        expect(topNode).toBeDefined();
        expect(topNode.text()).toBe('Label');
    });

    it('should render with `placeholder` from props', () => {
        const input = shallow(<Input placeholder='Placeholder' />);
        const controlNode = input.find('input');

        expect(controlNode.props().placeholder).toBe('Placeholder');
    });

    it('should render with `hint` from props', () => {
        const input = shallow(<Input hint='Hint' />);
        const subNode = input.find('.input__sub');

        expect(subNode).toBeDefined();
        expect(subNode.text()).toBe('Hint');
    });

    it('should render with `error` from props', () => {
        const input = shallow(<Input error='Error' />);
        const subNode = input.find('.input__sub');

        expect(subNode).toBeDefined();
        expect(subNode.text()).toBe('Error');
    });

    it('should render with `off` autocomplete attribute', () => {
        const input = shallow(<Input autocomplete={ false } />);
        const controlNode = input.find('input');

        expect(controlNode.props().autoComplete).toBe('off');
    });

    it('should render with `on` autocomplete attribute', () => {
        const input = shallow(<Input autocomplete={ true } />);
        const controlNode = input.find('input');

        expect(controlNode.props().autoComplete).toBe('on');
    });

    it('should render string autocomplete attribute', () => {
        const input = shallow(<Input autocomplete='email' />);
        const controlNode = input.find('input');

        expect(controlNode.props().autoComplete).toBe('email');
    });

    it('should render with `data-test-id` attribute if it is set', () => {
        const input = mount(<Input data-test-id='some value' />);
        const rootNode = input.getDOMNode();

        expect(rootNode.getAttribute('data-test-id')).toBe('some value');
    });

    it('should set value from props', () => {
        const input = shallow(<Input value='text' />);
        const controlNode = input.find('input');

        expect(controlNode.props().value).toBe('text');
    });

    it('should change the state value when props is not defined', () => {
        const input = mount(<Input />);
        const controlNode = input.find('input');

        controlNode.simulate('change', { target: { value: 'other value' } });

        expect(input.find('input').props().value).toBe('other value');
    });

    it('should call `onFocus` callback after input was focused', () => {
        const onFocus = jest.fn();
        const input = mount(<Input onFocus={ onFocus } />);

        input.find('input').simulate('focus');

        expect(onFocus).toHaveBeenCalled();
    });

    it('should call `onClick` callback after input was clicked', () => {
        const onClick = jest.fn();
        const input = shallow(<Input onClick={ onClick } />);
        const controlNode = input.find('input');

        controlNode.simulate('click');

        expect(onClick).toHaveBeenCalled();
    });

    it('should call `onKeyDown` callback after key down in input', () => {
        const onKeyDown = jest.fn();
        const input = shallow(<Input onKeyDown={ onKeyDown } />);
        const controlNode = input.find('input');

        controlNode.simulate('keyDown');

        expect(onKeyDown).toHaveBeenCalled();
    });

    it('should call `onBlur` callback after input was blured', () => {
        const onBlur = jest.fn();
        const input = mount(<Input onBlur={ onBlur } />);

        input.find('input').simulate('blur');

        expect(onBlur).toHaveBeenCalled();
    });

    it('should call `onChange` callback', () => {
        const onChange = jest.fn();
        const input = shallow(<Input onChange={ onChange } />);
        const controlNode = input.find('input');

        controlNode.simulate('change', { target: { value: 'other value' } });

        expect(onChange).toHaveBeenCalled();
    });

    it('should render with clear element', () => {
        const input = shallow(<Input clear={ true } value='text' />);
        const clearNode = input.find('.input__clear');

        expect(clearNode.length).toBe(1);
    });

    it('should call `onClearClick` callback on clear icon click', () => {
        const onClearClick = jest.fn();
        const input = mount(<Input clear={ true } onClearClick={ onClearClick } />);
        const controlNode = input.find('input');

        controlNode.simulate('change', { target: { value: 'value' } });
        const clearNode = input.find('.input__clear').at(0);

        clearNode.simulate('click');

        expect(onClearClick).toHaveBeenCalled();
    });

    it('should clear value after clear icon click', () => {
        const input = mount(<Input clear={ true } />);
        const controlNode = input.find('input');

        controlNode.simulate('change', { target: { value: 'value' } });
        const clearNode = input.find('.input__clear').at(0);

        clearNode.simulate('click');

        expect(controlNode.props().value).toBe('');
    });

    it('should call `onChange` listener after clear icon click', () => {
        const onChange = jest.fn();
        const input = mount(
            <Input
                value='text'
                clear={ true }
                onChange={ onChange }
            />
        );
        const clearNode = input.find('.input__clear').at(0);

        clearNode.simulate('click');

        expect(onChange).toHaveBeenCalled();
    });

    it('should render with icon', () => {
        const input = shallow(<Input icon={ <IconSearch /> } />);
        const iconNode = input.find('.input__icon');

        expect(iconNode.length).toBe(1);
    });

    it('should return root `HTMLElement` after `getNode` method call', () => {
        const input = mount<Input>(<Input />);

        const node = input.instance().getNode();

        expect(node).toBeInstanceOf(HTMLElement);
        expect(node).toBe(input.getDOMNode());
    });

    // it('should format masked value with custom formatter', () => {
    //     const cyrillic = {
    //         c: {
    //             validate(char) { return /^[ЁёА-Яа-я]$/.it(char); }
    //         }
    //     };
    //     let input = mount(
    //         <Input maskFormatCharacters={ cyrillic } mask='c 111 cc' value='12-3hjhkА456ИТг' />
    //     );
    //     let controlNode = input.find('input');
    //
    //     expect(controlNode.value).toBe('А 456 ИТ');
    // });

    it(
        'should return `HTMLInputElement` when `getControl` method called',
        () => {
            const input = mount<Input>(<Input />);
            const controlNode = input.instance().getControl();

            expect(controlNode).toBeInstanceOf(HTMLInputElement);
        }
    );

    it(
        'should return `HTMLInputElement` when `getControl` method called and mask is set',
        () => {
            const input = mount<Input>(<Input mask='111' />);
            const controlNode = input.instance().getControl();

            expect(controlNode).toBeInstanceOf(HTMLInputElement);
        }
    );

    it(
        'should return null when `getMaskedInputInstance` method is called and mask is not set',
        () => {
            const input = shallow<Input>(<Input />);
            const maskedInputInstance = input.instance().getMaskedInputInstance();

            expect(maskedInputInstance).toBeNull();
        }
    );

    it(
        'should return MaskedInput instance when getMaskedInputInstance method is called and mask is set',
        () => {
            const input = mount<Input>(<Input mask='111' />);
            const maskedInputInstance = input.instance().getMaskedInputInstance();

            expect(maskedInputInstance).toBeInstanceOf(MaskedInput);
        }
    );

    it('should call `onTouchStart` callback after input was touched', () => {
        const onTouchStart = jest.fn();
        const input = shallow(<Input onTouchStart={ onTouchStart } />);
        const controlNode = input.find('input');

        controlNode.simulate('touchStart');

        expect(onTouchStart).toHaveBeenCalled();
    });

    it(
        'should call `onTouchEnd` callback after input was released on touch',
        () => {
            const onTouchEnd = jest.fn();
            const input = shallow(<Input onTouchEnd={ onTouchEnd } />);
            const controlNode = input.find('input');

            controlNode.simulate('touchEnd');

            expect(onTouchEnd).toHaveBeenCalled();
        }
    );

    it(
        'should call `onTouchCancel` callback after input was disrupted on touch',
        () => {
            const onTouchCancel = jest.fn();
            const input = shallow(<Input onTouchCancel={ onTouchCancel } />);
            const controlNode = input.find('input');

            controlNode.simulate('touchCancel');

            expect(onTouchCancel).toHaveBeenCalled();
        }
    );

    it(
        'should call `onTouchMove` callback after input was touched and moved',
        () => {
            const onTouchMove = jest.fn();
            const input = shallow(<Input onTouchMove={ onTouchMove } />);
            const controlNode = input.find('input');

            controlNode.simulate('touchMove');

            expect(onTouchMove).toHaveBeenCalled();
        }
    );

    it(
        'should call `onTouchStart` callback with argument as SyntheticEvent and type \'touchstart\'',
        () => {
            const onTouchStart = jest.fn();
            const input = mount(<Input onTouchStart={ onTouchStart } />);
            const controlNode = input.find('input');

            controlNode.simulate('touchStart');

            expect(onTouchStart).toHaveBeenCalled();
            expect(onTouchStart).toHaveBeenCalledWith(expect.objectContaining({ type: 'touchstart' }));
        }
    );

    it(
        'should call `onTouchEnd` callback with argument as SyntheticEvent and type \'touchend\'',
        () => {
            const onTouchEnd = jest.fn();
            const input = mount(<Input onTouchEnd={ onTouchEnd } />);
            const controlNode = input.find('input');

            controlNode.simulate('touchEnd');
            expect(onTouchEnd).toHaveBeenCalled();
            expect(onTouchEnd).toHaveBeenCalledWith(expect.objectContaining({ type: 'touchend' }));
        }
    );

    it(
        'should call `onTouchMove` callback with argument as SyntheticEvent and type \'touchmove\'',
        () => {
            const onTouchMove = jest.fn();
            const input = mount(<Input onTouchMove={ onTouchMove } />);
            const controlNode = input.find('input');

            controlNode.simulate('touchMove');
            expect(onTouchMove).toHaveBeenCalled();
            expect(onTouchMove).toHaveBeenCalledWith(expect.objectContaining({ type: 'touchmove' }));
        }
    );

    it(
        'should call `onTouchCancel` callback with argument as SyntheticEvent and type \'touchcancel\'',
        () => {
            const onTouchCancel = jest.fn();
            const input = mount(<Input onTouchCancel={ onTouchCancel } />);
            const controlNode = input.find('input');

            controlNode.simulate('touchCancel');
            expect(onTouchCancel).toHaveBeenCalled();
            expect(onTouchCancel).toHaveBeenCalledWith(expect.objectContaining({ type: 'touchcancel' }));
        }
    );

    it(
        'should call `enableMouseWheel` and `disableMouseWheel` after focus and blur',
        () => {
            const input = mount<Input>(<Input />);
            const controlNode = input.find('input');

            jest.spyOn(input.instance(), 'enableMouseWheel');
            jest.spyOn(input.instance(), 'disableMouseWheel');

            controlNode.simulate('focus');
            expect(input.instance().enableMouseWheel).toHaveBeenCalled();

            controlNode.simulate('blur');
            expect(input.instance().disableMouseWheel).toHaveBeenCalled();
        }
    );

    it(
        'should call `resetError` after focus',
        () => {
            const input = mount<Input>(<Input />);
            const controlNode = input.find('input');
            // @ts-ignore
            jest.spyOn(input.instance(), 'resetError');

            controlNode.simulate('focus');
            // @ts-ignore
            expect(input.instance().resetError).toHaveBeenCalled();
        }
    );
});
