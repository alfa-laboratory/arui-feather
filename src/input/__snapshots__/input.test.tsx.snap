// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input should render without problems 1`] = `
<span
  className="input input_type_text input_view_default input_size_m"
>
  <span
    className="input__inner"
  >
    <span
      className="input__box"
      key="input-wrapper"
    >
      <span
        className="input__input-wrapper"
      >
        <input
          autoComplete="on"
          className="input__control"
          formNoValidate={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onPaste={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          type="text"
          value=""
          view="default"
        />
      </span>
    </span>
  </span>
</span>
`;
